### PR TITLE
[Khải Thư] Feature/search

### PR DESCRIPTION
**Pull request by dangcongkhaithu**

### CODE REVIEW

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import com.khaithu.a3ctech_android_final_project.adapter.SearchRecyclerAdapter
+import com.khaithu.a3ctech_android_final_project.adapter.interfaceadapter.IHandleEvent
+import com.khaithu.a3ctech_android_final_project.helper.Constant
+import com.khaithu.a3ctech_android_final_project.model.ResultMovie
+import com.khaithu.a3ctech_android_final_project.presenter.SearchPresenter
+import com.khaithu.a3ctech_android_final_project.view.MovieDetailActivity
+import com.khaithu.a3ctech_android_final_project.view.interfaceView.ISearchView
+import kotlinx.android.synthetic.main.fragment_search_view.view.*
+
+class SearchMovieView : Fragment(), ISearchView, IHandleEvent, HandlerPageEvent {
```
Comment (vuphu) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt]:
```suggestion
class SearchMovieFragment : Fragment(), ISearchView, IHandleEvent, HandlerPageEvent {
```

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/SearchPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                override fun onSubscribe(d: Disposable?) {
+                    if (d != null) {
+                        compositeDisposable.add(d)
+                    }
+                }
+
+                override fun onNext(t: Search?) {
+                    if (t != null) {
+                        mISearchView.getTotalPage(t.totalPage)
+                        mISearchView.onSearchReceiver(t.results)
```
Comment (vuphu) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/SearchPresenter.kt]: combine two lines
```suggestion
                        mISearchView.onSearchSuccess(t.results, t.totalPage)
```

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/SearchPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import com.khaithu.a3ctech_android_final_project.service.IMovieService
+import com.khaithu.a3ctech_android_final_project.view.interfaceView.ISearchView
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
+import io.reactivex.rxjava3.core.Observable
+import io.reactivex.rxjava3.core.Observer
+import io.reactivex.rxjava3.disposables.CompositeDisposable
+import io.reactivex.rxjava3.disposables.Disposable
+import io.reactivex.rxjava3.schedulers.Schedulers
+
+class SearchPresenter(val mISearchView: ISearchView) : BasePresenter(), ISearchPresenter {
```
Comment (vuphu) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/SearchPresenter.kt]:
```suggestion
class SearchPresenter(var mView: ISearchView?) : BasePresenter(), ISearchPresenter {
```
assign `mView = null` when destroy presenter

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+class PageAdapter(var list: MutableList<Int>, var mHandlerPageEvent: HandlerPageEvent) :
+    RecyclerView.Adapter<PageAdapter.PageNumberViewHolder>() {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PageNumberViewHolder {
+        val inflater: LayoutInflater = LayoutInflater.from(parent.context)
+        val view: View = inflater.inflate(R.layout.item_pagination_number, parent, false)
+        return PageNumberViewHolder(view, mHandlerPageEvent)
+    }
+
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt:21]: remove

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    override fun getItemCount(): Int {
+        return list.size
+    }
+
+    override fun onBindViewHolder(holder: PageNumberViewHolder, position: Int) {
+        val pageHolder: PageNumberViewHolder = holder as PageNumberViewHolder
+        pageHolder.bind(list[position] as Int)
+
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt:30]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/SearchRecyclerAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import kotlinx.android.synthetic.main.item_search.view.*
+
+class SearchRecyclerAdapter(var mHandleEvent: IHandleEvent) :
+    RecyclerView.Adapter<SearchRecyclerAdapter.ViewHolder>() {
+    private var lists: List<ResultMovie> = emptyList()
+
+    fun updateData(movies: List<ResultMovie>) {
+        lists = movies
+        notifyDataSetChanged()
+    }
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/SearchRecyclerAdapter.kt]: move below override

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/SearchPresenter.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    private val movieService: IMovieService = RestClient.getMovieService()
+    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
+    private var mPage: Int = 1
+
+    fun updatePage(page: Int) {
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/presenter/SearchPresenter.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    override fun onClickEvent(id: Int) {
+        val intent = Intent(context, MovieDetailActivity::class.java)
+        intent.putExtra(Constant.intentDetailMovie, id)
+        startActivity(intent)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        mPresenter.onDestroyPresenter()
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt]: move up

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            override fun onQueryTextChange(newText: String?): Boolean {
+                return false
+            }
+        })
+        return mView
+    }
+
+    override fun onSearchReceiver(results: List<ResultMovie>) {
+        mSearchAdapter.updateData(results)
+        onCreatePaginationRecyclerView(mView, 80)
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt]: define 80

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        mView = inflater.inflate(R.layout.fragment_search_view, container, false)
+
+        val searchView = mView.search
+
+        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt]: No need to define searchView variable

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            override fun onQueryTextSubmit(query: String?): Boolean {
+                if (query != null) {
+                    mQuery = query
+                    mPresenter.search(mQuery!!)
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt]: do not force not null

File: app/src/main/res/layout/item_pagination_number.xml [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content">
+
+    <TextView
+        android:id="@+id/textNumber"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:textColor="@color/black"
+        android:text="5"
```
Comment (AnhJobChat) [inline app/src/main/res/layout/item_pagination_number.xml]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        mSearchAdapter = SearchRecyclerAdapter(this)
+        view.searchRecycler.adapter = mSearchAdapter
+        view.searchRecycler.layoutManager =
+            LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
+    }
+
+    private fun onCreatePaginationRecyclerView(view: View, totalPage: Int) {
+        for (i in 1..totalPage) {
+                mList.add(i)
+        }
```
Comment (AnhJobChat) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieView.kt]: format?

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,41 @@
+package com.khaithu.a3ctech_android_final_project.adapter
+
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import com.khaithu.a3ctech_android_final_project.R
+import kotlinx.android.synthetic.main.item_pagination_number.view.*
+
+class PageAdapter(var list: MutableList<Int>, var mHandlerPageEvent: HandlerPageEvent) :
```
Comment (chuongdevJC) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt]: avoid pass params in adapter constructor if its obligatory. should make a function update better. we do not dependent on constructor.

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/SearchRecyclerAdapter.kt [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import com.khaithu.a3ctech_android_final_project.R
+import com.khaithu.a3ctech_android_final_project.adapter.interfaceadapter.IHandleEvent
+import com.khaithu.a3ctech_android_final_project.helper.Constant
+import com.khaithu.a3ctech_android_final_project.helper.GlideHelper
+import com.khaithu.a3ctech_android_final_project.model.ResultMovie
+import kotlinx.android.synthetic.main.item_search.view.*
+
+class SearchRecyclerAdapter(var mHandleEvent: IHandleEvent) :
```
Comment (chuongdevJC) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/SearchRecyclerAdapter.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieFragment.kt [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        onCreatePaginationRecyclerView(mView, totalPage)
+    }
+
+    override fun onClickEvent(id: Int) {
+        val intent = Intent(context, MovieDetailActivity::class.java)
+        intent.putExtra(Constant.intentDetailMovie, id)
+        startActivity(intent)
+    }
+
+    override fun onDestroy() {
```
Comment (chuongdevJC) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/view/fragment/SearchMovieFragment.kt:67]: remember destroy all listener that we defined

**Review by vuphu-jc — APPROVED** [review]

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    override fun getItemCount(): Int {
+        return list.size
+    }
+
+    override fun onBindViewHolder(holder: PageNumberViewHolder, position: Int) {
+        val pageHolder: PageNumberViewHolder = holder
+        pageHolder.bind(list[position])
+    }
+
+     fun updateData(handlerPageEvent: HandlerPageEvent, list: MutableList<Int>) {
```
Comment (chuongdevJC) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt]: .

File: app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import com.khaithu.a3ctech_android_final_project.R
+import kotlinx.android.synthetic.main.item_pagination_number.view.*
+
+class PageAdapter :
+    RecyclerView.Adapter<PageAdapter.PageNumberViewHolder>() {
+
+    private lateinit var mHandlerPageEvent: HandlerPageEvent
+    private lateinit var list: MutableList<Int>
```
Comment (chuongdevJC) [inline app/src/main/java/com/khaithu/a3ctech_android_final_project/adapter/PageAdapter.kt]: do not set lateinit when we do not sure these variable will be init later.

**Review by AnhJobChat — APPROVED** [review]

**Review by chuongdevJC — APPROVED** [review]

**Pull request merged**